presto-main/src/main/java/com/facebook/presto/operator/ThriftRpcShuffleClient.java

```diff
@@ -0,0 +1,95 @@
+ */
+package com.facebook.presto.operator;
+
+import com.facebook.airlift.log.Logger;
+import com.facebook.drift.client.DriftClient;
+import com.facebook.presto.execution.TaskId;
+import com.facebook.presto.execution.buffer.BufferResult;
+import com.facebook.presto.execution.buffer.OutputBuffers.OutputBufferId;
+import com.facebook.presto.operator.PageBufferClient.PagesResponse;
+import com.facebook.presto.server.thrift.ThriftTaskClient;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import io.airlift.units.DataSize;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.net.URI;
+import java.util.Optional;
+import java.util.concurrent.Executor;
+
+import static com.facebook.presto.operator.PageBufferClient.PagesResponse.createPagesResponse;
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+public final class ThriftRpcShuffleClient
+        implements RpcShuffleClient
+{
+    private static final Logger log = Logger.get(ThriftRpcShuffleClient.class);
+
+    private final ThriftTaskClient thriftClient;
+    private final Executor executor;
+    private final TaskId taskId;
+    private final OutputBufferId outputBufferId;
+
+    public ThriftRpcShuffleClient(DriftClient<ThriftTaskClient> driftClient, URI location, Executor executor)
+    {
+        requireNonNull(location, "location is null");
+
+        this.thriftClient = requireNonNull(driftClient, "thriftClient is null").get(Optional.of(location.getAuthority()));
+        this.executor = requireNonNull(executor, "executor is null");
+
+        // TODO: refactor the entire LocationFactory interfaces to completely replace URI with more efficient/expressive data structures
+        // location format: thrift://{ipAddress}:{thriftPort}/v1/task/{taskId}/results/{bufferId}/
+        String[] paths = location.getPath().split("/");
+        this.taskId = TaskId.valueOf(paths[3]);
+        this.outputBufferId = OutputBufferId.fromString(paths[5]);
+    }
+
+    @Override
+    public ListenableFuture<PagesResponse> getResults(long token, DataSize maxResponseSize)
+    {
+        ListenableFuture<BufferResult> future = thriftClient.getResults(taskId, outputBufferId, token, maxResponseSize.toBytes());
+        return Futures.transform(
+                future,
+                result -> createPagesResponse(
+                        result.getTaskInstanceId(),
+                        result.getToken(),
+                        result.getNextToken(),
+                        result.getSerializedPages(),
+                        result.isBufferComplete()),
+                directExecutor());
+    }
+
+    @Override
+    public void acknowledgeResultsAsync(long nextToken)
+    {
+        executor.execute(() -> {
+            try {
+                thriftClient.acknowledgeResults(taskId, outputBufferId, nextToken);
```

Comment:
Instead of popping this into an executor, can we instead just make the Thrift interface async?  We don't need to handle the future if we don't care about the result.

Reply:
Interesting. We could avoid using the executor in that case as well.